Add jira reference support

randompixel opened the pull request:
## Description
This pull request adds the ability to:
 - Set a issue tracker format in the settings (defaults to github)
 - Validate JIRA issue keys if issue tracker format is set to JIRA
 - Remove the preceeding # if issue tracker format is set to JIRA

Go easy on me, it is my first attempt at node. I was thinking about a larger refactoring where the issue tracker is an object that encapsulates it's own rules and formatting, so like
```
class JIRAFormatter {
    function getPrefix() {};
    function isValid() {};
    function getCommitTemplate() { // {issuekey} {summary} - {description} }
} 

class GithubFormatter {
    function getPrefix() {};
    function isValid() {}
    function getCommitTemplate() { // {summary} - {description} {issuekey} }
}
```
... but I decided against it for the time being! Maybe another day.

All feedback welcome. :)

Issue: #44 

## Tests

Not sure how to run the tests - help welcome!

randompixel commented:
<img width="544" alt="screen shot 2017-05-03 at 15 03 17"
 src="https://cloud.githubusercontent.com/assets/2316601/25664184/c43d7716-3011-11e7-9157-ae78e03795bc.png">

 _^^ example within Bitbucket Cloud_

randompixel commented:
I will fix the build issues tonight and push again.

carloscuesta commented:
Hello @randompixel and thanks for sending the pull request!

Before reviewing, let's fix the build :white_check_mark:. To run the tests locally you just need to `npm install` and `npm run test`. In fact you can take a look at the scripts of the project at [package.json scripts](https://github.com/carloscuesta/gitmoji-cli/blob/master/package.json#L21)

As you can see at the [Travis build](https://travis-ci.org/carloscuesta/gitmoji-cli/jobs/228351077#L230) there are style errors in the code.

```
src/gitmoji.js:184:21
  ✖  184:21  Expected indentation of 5 tabs but found 20 spaces.  indent
  ✖  185:7   Expected indentation of 1 tab but found 6.           indent
  ✖  186:6   Expected indentation of 0 tabs but found 5.          indent
  ✖  188:25  Expected indentation of 6 tabs but found 24 spaces.  indent
  ✖  189:25  Expected indentation of 6 tabs but found 24 spaces.  indent
  ✖  194:8   Expected indentation of 8 tabs but found 7.          indent
  ✖  200:25  Expected indentation of 6 tabs but found 24 spaces.  indent
  ✖  201:29  Expected indentation of 1 tab but found 28 spaces.   indent
  ✖  202:25  Expected indentation of 0 tabs but found 24 spaces.  indent
  ✖  203:25  Expected indentation of 6 tabs but found 24 spaces.  indent
```

randompixel commented:
Hey @carloscuesta,

I've fixed the code style issues and split out the commit tests to cater for the different formats. Travis seems happier now :)

As an aside, I don't this is the nicest way of testing it as I think it will change the user preferences to JIRA every time if they run the tests locally. Not an issue day-to-day or in the CI environment, however without refactoring more of the code to pass configuration objects around I'm not sure how to get around this.

carloscuesta commented on src/gitmoji.js:
```diff
@@ -116,7 +126,8 @@ class GitmojiCli {
 	_commit(answers) {
 		const title = `${answers.gitmoji} ${answers.title}`;
-		const reference = (answers.reference) ? `#${answers.reference}` : '';
+		const prefix = config.get('issueFormat') === 'github' ? '#' : '';
```
I think it's more clear to rename `prefix` to prefixReference

randompixel replied:
ok, will change that.

carloscuesta commented on src/gitmoji.js:
```diff
@@ -168,17 +179,28 @@ class GitmojiCli {
+						let errorReference = '';
+						switch (config.get('issueFormat')) {
+							case 'jira':
+								validReference = value.match(/^([a-zA-Z]{1,10}-[0-9]+)$/g);
```
Never worked with Jira, are you sure that regular expression follows the issue format ? 

As I'm seeing in [Jira Docs](https://confluence.atlassian.com/jirasoftwarecloud/linking-issues-776997756.html) the pattern is like `XXX-11111111` three letters and no limit for the number, is that correct?

randompixel replied:
By default it is three letters and unlimited numbers. However, it is configurable. Based on the [docs here](https://confluence.atlassian.com/adminjiraserver073/changing-the-project-key-format-861253229.html) it suggests `([A-Z][A-Z]+)` which is much looser but I will change it to that.

carloscuesta replied:
No problem on that, but let's stick with the default case for Jira!

randompixel commented:
OK, I've rebased and updated based on your suggestions. 

I've gone with a regexp that meets this requirement from the project creation page - the key is the first part of the issue reference before the hyphen. Anything after the hyphen should just be a number.
![image](https://cloud.githubusercontent.com/assets/2316601/25839396/5ead421e-348f-11e7-8183-93248f8c258d.png)

carloscuesta commented:
Thanks! @randompixel

randompixel commented:
Thanks for the review @carloscuesta - much appreciated :)
